File: implementations/php/pcf-dcp/src/FragmentEntry.php
<?php

declare(strict_types=1);

namespace Kduma\PCFDCP;

/** One Fragment Entry: a single extent of an inner partition (spec 8.2). */
final class FragmentEntry
{
    public function __construct(
        public int $extentOffset,
        public int $extentLength,
        public int $kind,
        public int $flags,
    ) {
    }

    /** Serialise to the on-disk 18-byte layout. */
    public function toBytes(): string
    {
        return pack('P', $this->extentOffset)
            . pack('P', $this->extentLength)
            . \chr($this->kind & 0xFF)
            . \chr($this->flags & 0xFF);
    }

    /** Parse from the on-disk 18-byte layout (optionally at an offset). */
    public static function fromBytes(string $b, int $offset = 0): self
    {
        return new self(
            unpack('P', substr($b, $offset + 0, 8))[1],
            unpack('P', substr($b, $offset + 8, 8))[1],
            \ord($b[$offset + 16]),
            \ord($b[$offset + 17]),
        );
    }

    /** Whether this entry's kind is DATA. */
    public function isData(): bool
    {
        return $this->kind === Consts::KIND_DATA;
    }

    /** Whether the SHARED flag (bit 0) is set. */
    public function isShared(): bool
    {
        return ($this->flags & Consts::FLAG_SHARED) !== 0;
    }
}

File: implementations/dotnet/pcf-dcp/src/Pcf.Dcp/FragmentEntry.cs
namespace Pcf.Dcp;

/// <summary>One Fragment Entry: a single extent of an inner partition (spec 8.2).</summary>
public sealed class FragmentEntry
{
    /// <summary>Arena-relative start of the extent's bytes.</summary>
    public ulong ExtentOffset { get; set; }

    /// <summary>Length of the extent in bytes.</summary>
    public ulong ExtentLength { get; set; }

    /// <summary>Extent kind (<c>1</c> = DATA; <c>0</c> invalid; <c>2</c>/<c>3</c> reserved).</summary>
    public byte Kind { get; set; }

    /// <summary><c>flags</c> byte (bit 0 = SHARED; others reserved 0).</summary>
    public byte Flags { get; set; }

    /// <summary>Serialise to the on-disk 18-byte layout.</summary>
    public byte[] ToBytes()
    {
        var b = new byte[Constants.FragmentEntrySize];
        LittleEndian.WriteU64(b, 0, ExtentOffset);
        LittleEndian.WriteU64(b, 8, ExtentLength);
        b[16] = Kind;
        b[17] = Flags;
        return b;
    }

    /// <summary>Parse from the on-disk 18-byte layout.</summary>
    public static FragmentEntry FromBytes(byte[] b, int offset = 0)
    {
        return new FragmentEntry
        {
            ExtentOffset = LittleEndian.ReadU64(b, offset + 0),
            ExtentLength = LittleEndian.ReadU64(b, offset + 8),
            Kind = b[offset + 16],
            Flags = b[offset + 17],
        };
    }

    /// <summary>Whether this entry's <c>kind</c> is DATA.</summary>
    public bool IsData() => Kind == Constants.KindData;

    /// <summary>Whether the SHARED flag (bit 0) is set.</summary>
    public bool IsShared() => (Flags & Constants.FlagShared) != 0;
}
